File: leetcode/1029.two-city-scheduling.py
# @leet start
class Solution:
    def twoCitySchedCost(self, costs: list[list[int]]) -> int:
        # Sort the savings in descending order of savings
        costs.sort(key=lambda l: abs(l[0] - l[1]), reverse=True)

        # Greedily assign the person with the current largest savings to
        # their cheaper city if possible
        n = len(costs) // 2

        city_a_count = 0
        city_b_count = 0

        total_cost = 0

        for city_a_cost, city_b_cost in costs:
            # Assign to city A if it's cheaper and it's not full or if
            # city B is full; else assign to city B
            if city_a_cost <= city_b_cost and city_a_count < n or city_b_count == n:
                total_cost += city_a_cost
                city_a_count += 1
            else:
                total_cost += city_b_cost
                city_b_count += 1

        return total_cost


# @leet end

File: leetcode/1029.two-city-scheduling.js
// @leet start
/**
 * @param {number[][]} costs
 * @return {number}
 */
var twoCitySchedCost = function (costs) {
  return costs
    .sort((a, b) => a[0] - a[1] - (b[0] - b[1]))
    .reduce((acc, curr, i) => acc + curr[i < costs.length / 2 ? 0 : 1], 0);
};
// @leet end
